### ISSUE TITLE
Content-Length not checked

### ISSUE THREAD

**Issue by acornea** [clearlinux/swupd-client#44]
A malicious/misbehaving server can respond without Content-Length header data and send infinite data to requests, which will block the client. 
Tested on swupd v2.87 with a custom server.

Below the HTTPRequestHandler for a python HTTP Server for reproduction:

```
class InfiniteData(BaseHTTPServer.BaseHTTPRequestHandler):
    """Handler that returns infinite data"""
    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        while True:
            self.wfile.write(chr(255))
```

One solution would be to actual check for Content-Length header information.

**Comment by phmccarty** [comment]
I agree that we should check for the Content-Length header. If it is not present, then we should be able to check CURLINFO_CONTENT_LENGTH_DOWNLOAD before starting the download.

**Comment by matthewrsj** [comment]
We need to decide what should be done in the event that content length is not set. Do we really want to fail on that condition?

**Comment by matthewrsj** [comment]
@icarus-sparry I believe you mentioned we can't trust the content-length header, or they were irrelevant for HTTP2, I forget which. It has been a long time since we talked about it. Can you remind me here?

**Comment by icarus-sparry** [comment]
We can't trust Content-length will be set. If it is set then libcurl will honor it for potentially multiplexed connections (http 2.0, http 1.1).

This is why we want the size of the download to be specified in the manifest. Then the only DOS is on the download of the MoM, and we can bound the upper size of that.

**Comment by matthewrsj** [comment]
@icarus-sparry see my request in #101

**Comment by icarus-sparry** [comment]
#403 created in response to request in #101

**Comment by matthewrsj** [comment]
Thanks @icarus-sparry

**Comment by cmarcelo** [comment]
We do have an _upper bound_ for those downloads: the contentsize in each manifest, so except for MoM we have a solution for the unbounded downloads.

- Bundle Manifests download should be capped by the contentsize in Manifest.MoM (this is currently zero, but could be set with the sum of all manifests).
- Files downloads should be capped by the contentsize in the bundle they are referred to.
- Packs downloads should be capped by the contentsize of their respective bundle. In theory our compression is bad and we would end up with only non compressed files, so we should account for some tar overhead based on the number of files involved.

Some upper bounds are really larger than the real files sizes, but moves us from "never ends" to "might take a long time but finishes", without depending on larger changes.

For MoM, I'm OK with capping to some arbitrary big file.

Another possibility (for MoM and even other bundles) could be: start parsing it as we download it, so we can detect something invalid and bail early (and even bail if we see valid but unreasonable stuff like filecount larger than some number, record line larger than some number).

**Comment by cmarcelo** [comment]
I think this can be fixed independently of #403, so I'd consider reopening this issue. More data in Manifest can get us better upper bounds for downloads, but there is work that can be done here in parallel of that proposal: e.g. asking curl to cap the downloads at a certain upper bound.
